<?xml version="1.0" encoding="us-ascii"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
      "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
  <meta http-equiv="content-type" content="text/html; charset=us-ascii" />
  <title>Ancora technical documentation</title>
  <meta name="generator" content="Amaya, see http://www.w3.org/Amaya/" />
</head>

<body>
<h1>Ancora technical documentation</h1>

<p>First version by P&auml;r Engstr&ouml;m, April 2008. This document was
produced using the <a href="http://www.w3.org/Amaya/">Amaya HTML editor</a>.
Please use the same or some other editor that produces clean HTML for editing
this document.</p>

<p>This document describes the implementation the Ancora web resource. For a
general description of Ancora and its user interface, see the <a
href="http://genomebiology.com/2008/9/2/R34">Ancora publication</a>.</p>

<p id="Table"><strong>Table of contents</strong></p>

<div class="toc">
<ul>
  <li><a href="#Installati">1. Installation</a> 
    <ul>
      <li><a href="#Software">1.1. Software dependencies</a></li>
      <li><a href="#Location">1.2. Location of Ancora files</a></li>
      <li><a href="#Software1">1.3. Software components</a></li>
      <li><a href="#Setting">1.4. Setting up ancora-test</a></li>
      <li><a href="#Data">1.5. Data files</a></li>
      <li><a href="#Apache">1.6. Apache configuration</a></li>
      <li><a href="#database">1.7. The CNE database</a></li>
      <li><a href="#L1473">1.8. MySQL account for Perl scripts</a></li>
    </ul>
  </li>
  <li><a href="#Generating">2. Generating CNEs</a> 
    <ul>
      <li><a href="#Obtaining">2.1. Obtaining alignment files</a></li>
      <li><a href="#Creating">2.2. Creating filter files</a></li>
      <li><a href="#Scanning">2.3. Scanning for CNEs</a></li>
      <li><a href="#Removing">2.4. Removing unannotated repeats with
      BLAT</a></li>
      <li><a href="#Loading">2.5. Loading CNEs into the cne database</a></li>
    </ul>
  </li>
  <li><a href="#Setting1">3. Setting up a genome browser</a> 
    <ul>
      <li><a href="#Creating1">3.1. Creating an annotation database for
        GBrowse</a></li>
      <li><a href="#Obtaining1">3.2. Obtaining genome annotations</a></li>
      <li><a href="#Creating2">3.3. Creating and loading GFF files</a></li>
      <li><a href="#Adding">3.4. Adding information about the assembly to the
        CNE database</a></li>
      <li><a href="#Creating3">3.5. Creating a configuration file for
        GBrowse</a></li>
      <li><a href="#Making">3.6. Making download files</a></li>
    </ul>
  </li>
  <li><a href="#HTML">4. The HTML pages and CGI scripts</a></li>
  <li><a href="#service">5. The DAS service</a></li>
</ul>
</div>

<h2 id="Installati">1. Installation</h2>

<p>This document focuses on the current main Ancora installation on
shire.bccs.uib.no (running CentOS 5) but is also intended to be helpful for
setting up Ancora on other machines. Although only tested on CentOS, it
should not be difficult to get Ancora working under other Linux/Unix
operating systems.</p>

<h3 id="Software">1.1. Software dependencies</h3>

<p>The following software must be installed on the system:</p>
<ul>
  <li>A httpd server (we currently use <a
    href="http://www.apache.org/">Apache</a> version 2.2.3)</li>
  <li><a href="http://www.mysql.org">MySQL server and client</a> (we
    currently use MySQL version 5.0.22)</li>
  <li><a href="http://www.bioperl.org">BioPerl</a></li>
  <li><a href="http://www.gmod.org/wiki/index.php/GBrowse">GBrowse</a></li>
  <li><a href="http://genome.ucsc.edu/FAQ/FAQlicense">UCSC Genome Browser
    source and utilities</a></li>
  <li><a
    href="http://www.sanger.ac.uk/Software/analysis/proserver/">ProServer DAS
    server</a></li>
</ul>

<p>BioPerl and GBrowse in turn have numerous dependencies. Of those, it is
particularly important that the GD Perl library and its underlying C library
are correctly installed, since GBrowse uses GD extensively.</p>

<p>The current installation of Ancora uses GBrowse installed from source
obtained from the stable CVS branch on September 5, 2007. We had to make a
few changes to the GBrowse source code to fix issues related to caching and
ordering of tracks. We have reported these changes to the GBrowse developers,
and they should be fixed in more recent versions of GBrowse. Updating the
Ancora installation to a more recent GBrowse version should be done in the
near future. We keep the GBrowse files in the following directories:</p>
<ul>
  <li>html files under /opt/www/ancora/html/gbrowse</li>
  <li>cgi-bin scripts under /opt/www/ancora/cgi-bin</li>
  <li>configuration files under /opt/www/ancora/conf/gbrowse.conf</li>
</ul>

<p>This can be achieved at installation by giving appropriate flags to the
GBrowse Makefile.PL script:</p>
<pre>perl Makefile.PL HTDOCS=/opt/www/ancora/html \
                 CGIBIN=/opt/www/ancora/cgi-bin \
                 CONF=/opt/www/ancora/conf</pre>

<p>The UCSC Genome Browser source code is needed to compile the C program for
detecting CNEs, which makes use of several library functions from the UCSC
source. The UCSC Genome Browser source also contains a collection of useful
utilities, some of which are needed at steps below. It is recommended to
compile all the UCSC utilities by following the instructions in the README
file in the UCSC source package. On shire, this currently boils down to:</p>
<pre>export MACHTYPE=x86_64
export MYSQLINC=/usr/include/mysql
export MYSQLLIBS= "-L/usr/lib64/mysql -lmysqlclient -lz -lcrypt -lnsl -lm -L/usr/lib64 -lssl -lcrypto"
mkdir ~/bin/$MACHTYPE
cd kent/src
(make clean)
make libs</pre>

<p>The ProServer DAS server is only required for the DAS functionality of
Ancora. See the section on the DAS service below for details.</p>

<h3 id="Location">1.2. Location of Ancora files</h3>

<p>The live installation of Ancora - accessible at <a
href="http://ancora.genereg.net/">http://ancora.genereg.net/</a> - is located
under /opt/www/ancora on shire.bccs.uib.no. There is also a directory
/opt/www/ancora-test that holds a development version accessible at <a
href="http://ancora-test.genereg.net/">http://ancora-test.genereg.net/</a>.</p>

<p>In addition, Ancora makes use of software libraries, flat files and MySQL
databases in other locations on the same machine. These are described in more
detail below and are located under the following directories:</p>
<ul>
  <li>Perl libraries: 
    <ul>
      <li>/opt/www/cne/perl-lib</li>
      <li>/opt/www/AT/lib</li>
    </ul>
  </li>
  <li>Genome assembly sequences in binary 2bit format: 
    <ul>
      <li>/export/data/goldenpath</li>
    </ul>
  </li>
  <li>MySQL databases: 
    <ul>
      <li>/var/lib/mysql/cne</li>
      <li>/var/lib/mysql/gbrowse_gff_*</li>
    </ul>
  </li>
</ul>

<p>Other directories contain files used in the generation of data for Ancora,
but that are not directly used by the web resource:</p>
<ul>
  <li>Scripts and source code for compiled programs: 
    <ul>
      <li>/opt/www/cne/scripts</li>
      <li>/opt/www/cne/tools</li>
      <li>/opt/www/cne/ancora/gbrowse_db</li>
      <li>/opt/www/AT/SCRIPTS</li>
    </ul>
  </li>
  <li>Whole-genome alignments: 
    <ul>
      <li>/export/downloads/ucsc/axtNet/</li>
    </ul>
  </li>
  <li>Files listing CNEs and filters: 
    <ul>
      <li>/export/data/CNEs</li>
    </ul>
  </li>
  <li>MySQL databases with genome annotation: 
    <ul>
      <li>/var/lib/mysql/UCSC_*</li>
    </ul>
  </li>
</ul>

<p>Installation and running of Ancora involves data files in several formats
(bed, axt, wig, 2bit) used by the UCSC Genome Browser database. For a
description of these formats, see the <a
href="http://genome.ucsc.edu/FAQ/FAQformat">UCSC Genome Bioinformatics
site</a>.</p>

<h3 id="Software1">1.3. Software components</h3>

<p>Ancora requires two software packages developed in Boris Lenhard&rsquo;s
group: cne and AT. The cne package contains Perl libraries, C source code and
Perl and bash scripts for detecting and working with CNEs, as well as Perl
libraries and CGI scripts directly underlying the Ancora web resource. The AT
package is a collection of Perl modules related to genome and transcriptome
sequence analysis. They were largely implemented by P&auml;r Engstr&ouml;m
during his PhD studies but also include contributions from several other past
and present members of the Lenhard group. Ancora only uses a few well-tested
modules from the AT package, which also includes many modules in early stages
of development.</p>

<p>The cne and AT packages are maintained as subversion repositories. To
retrieve the most recent versions, do:</p>
<pre>svn co http://www.ii.uib.no/svn/lenhard/AT
svn co http://www.ii.uib.no/svn/lenhard/cne</pre>

<p>One copy of the packages should be placed under /opt/www/cne and
/opt/www/AT, respectively, where the Ancora genome browser will access them
from. If you intend to make changes to the libraries, you may wish to have an
additional copy in your home directory (P&auml;r has them under ~/devel/AT
and ~/devel/cne). To run scripts that use the modules, you will have to add
the installation paths to your PERL5LIB environment variable. If you use bash
as your shell, you can do this by adding the following line to the file
~/.bash_profile:</p>
<pre>export PERL5LIB=/opt/www/AT/lib:/opt/www/cne/perl_lib:$PERL5LIB</pre>

<p>(If you want scripts to use a working copy of the modules in a different
location, modify the above line accordingly.)</p>

<p>The cne/tools directory in the cne package contains source code for
several C programs. Ancora currently only needs one of these programs:
ceScan, the program we use to detect CNE. Instructions for how to compile the
C programs are in cne/tools/README.txt.</p>

<p>The cne/ancora directory in the cne package contains files related to the
Ancora web resource. If you are setting up the Ancora web resource, you will
need to copy some files from this directory to the ancora directory on your
server. On our server shire this is /opt/www/ancora. The following commands
copy the files there:</p>
<pre>cd cne/ancora
cp -rp Bio-Das-ProServer/ancora /opt/www/ancora/Bio-Das-ProSever
cp -rp cgi-bin/* /opt/www/ancora/cgi-bin
cp -rp conf/gbrowse.conf/*.conf /opt/www/ancora/conf/gbrowse.conf
cp -rp conf/gbrowse.conf/plugins/* /opt/www/ancora/conf/gbrowse.conf/plugins
cp -rp html/downloads html/*.html html/*.css html/*.png /opt/www/ancora/html
cp -rp html/gbrowse/main.css /opt/www/ancora/html/gbrowse</pre>

<p>At some point, we should write an install script that takes care of
copying these files.</p>

<h3 id="Setting">1.4. Setting up ancora-test</h3>

<p>We currently keep a development version of ancora under
/opt/www/ancora-test on shire. This version can be accessed at <a
href="http://ancora-test.genereg.net/">http://ancora-test.genereg.net/</a>.
We are considering moving the development version to a different server.
Setting it up is currently a little awkward and should be simplified by a
configurable install script. First, copy the ancora files to
/opt/www/ancora-test (i.e. execute the copy commands in the previous section,
substituting /opt/www/ancora-test for /opt/www/ancora). Then make the
following changes to files under /opt/www/ancora-test:</p>
<ul>
  <li>(list changes here)</li>
</ul>

<h3 id="Data">1.5. Data files</h3>

<p>Ancora expects genome assembly sequences in binary 2bit format to be
present under /export/data/goldenpath. There should be one subdirectory for
each assembly, named using a UCSC assembly identifier (e.g. hg18, mm9) or
some other suitable identifier for assemblies not served by UCSC. Each such
subdirectory should contain a file named assembly.2bit containing the
soft-masked assembly sequence. Two-bit files can be made by downloading the
assembly in soft-masked fasta format from <a
href="http://genome.ucsc.edu/">http://genome.ucsc.edu/</a> and converting it
to 2bit format using the program faToTwoBit, which is part of the UCSC
utilities. E.g. for the latest human assembly:</p>
<pre>cd /export/data/goldenpath
mkdir hg18
cd hg18
wget http://hgdownload.cse.ucsc.edu/goldenPath/hg18/bigZips/chromFa.zip
unip chromFa.zip
faToTwoBit *.fa assembly.2bit
rm *.fa chromFa.zip</pre>

<p>We use additional files (annotation and aligment files) from UCSC in the
CNE detection pipeline and in constructing the other annotation tracks shown
in the genome browser. How to retrieve these files is described in the
corresponding sections below.</p>

<h3 id="Apache">1.6. Apache configuration</h3>

<p>The Apache httpd server on shire is configured to look for html files
under /var/www/html and CGI scripts under /var/www/cgi-bin. These directories
contain symbolic links pointing to the actual Ancora directories, set up
as:</p>
<pre>ln -s /opt/www/ancora/html /var/www/html/ancora
ln -s /opt/www/ancora/cgi/bin /var/www/cgi/bin/ancora</pre>

<p>(and similarly for ancora-test)</p>

<p>The mapping of the domains ancora.genereg.net to and
ancora-test.genereg.net to these directories is achieved by VirtualHost
entries in the Apache configuration file etc/httpd/conf/httpd/conf:</p>
<pre>&lt;VirtualHost *:80&gt;
ServerName ancora.genereg.net
ScriptAlias /cgi-bin /var/www/cgi-bin/ancora
DocumentRoot /var/www/html/ancora
# Reverse proxy directives
ProxyPass /das http://localhost:9000/das
ProxyPassReverse /das http://localhost:9000/das
&lt;/VirtualHost&gt;</pre>
<pre>&lt;VirtualHost *:80&gt;
ServerName ancora-test.genereg.net
ScriptAlias /cgi-bin /var/www/cgi-bin/ancora-test
DocumentRoot /var/www/html/ancora-test
&lt;/VirtualHost&gt;</pre>

<p>The reverse proxy directives cause requests to <a
href="http://ancora.genereg.net/das">http://ancora.genereg.net/das</a> to be
passed on to http://localhost:9000/das, where the DAS server listens.</p>

<h3 id="database">1.7. The CNE database</h3>

<p>Ancora reads information about genome assemblies and CNEs from a database
called cne served by the MySQL server on the same host. Programmatic access
to this database is provided by the Perl module CNE::DB in the cne
package.</p>

<p>If this database does not exist, create it:</p>
<pre>mysql -u username -p -e 'create database cne'
mysql -u username -p -e 'grant select on cne.* to nobody@localhost'</pre>

<p>(Replacing username with your MySQL username, or root if necessary.) The
second command gives user nobody read access to the database. This is
necessary because Ancora accesses the database as user nobody.</p>

<p>Presently there is only one table that must exist in the database. This
table is called assembly and holds information about the genome assemblies
for which comparisons are available in Ancora. The cne package contains a
file with SQL commands that create the table and fills it with sample
data:</p>
<pre>mysql -u username -p cne &lt; cne/scripts/cne_pipeline/create_assembly_table.sql</pre>

<h3 id="L1473">1.8. MySQL account for Perl scripts</h3>

<p>Generating CNEs and setting up the annotation database for Ancora involves
running several Perl scripts that need to access local MySQL databases. Some
of these scripts require that a MySQL username and password are specified in
a file called MyPerlVars.pm located in a directory mentioned in your PERL5LIB
environment variable. I have such a file in a directory called .perl under my
home directory, and have added the following line to my .bash_profile to make
Perl look for libraries in this directory:</p>
<pre>export PERL5LIB=~/.perl:$PERL5LIB</pre>

<p>The file itself should contain the following lines:</p>
<pre>package MyPerlVars;
use warnings;
use strict;

use Exporter;

our @ISA = qw(Exporter);
our @EXPORT_OK = qw($sqlUser $sqlPass);

our $sqlUser = "username"; # replace username with your username
our $sqlPass = "password"; # replace password with your password

1;</pre>

<p>For security reasons, it is a good idea to change the permissions of
MyPerlVars.pm file so that nobody but you can read it.</p>

<h2 id="Generating">2. Generating CNEs</h2>

<p>To scan for CNEs, you need to have the cne package, the AT package and
genome assembly sequences installed as outlined above. In addition, you need
to obtain alignment files for the genomes that are to be compared, and create
filter files that specify which regions (typically exons and repeats) that
are to be ignored in the scanning. The alignment and filter files are not
required to run the web resource, so they can be removed after CNE
generation.</p>

<h3 id="Obtaining">2.1. Obtaining alignment files</h3>

<p>The program that scans for CNEs (ceScan) takes alignments in axt format as
input. We typically obtain alignment files in axt format from UCSC and keep
them under /export/downloads/ucsc/axtNet. This directory contains one
subdirectory for each genome assembly. The directories are named using UCSC
assembly identifiers, as for the genome assembly directories (see above).
Each directory contains axt files for pairwise net alignments with the
corresponding assembly as the reference; e.g. directory
/export/downloads/ucsc/axtNet/hg18 only contains net alignments where hg18 is
reference. To detect CNEs between two genomes, you need two sets of net
alignments - one where each genome is the reference (see the <a
href="http://genomebiology.com/2008/9/2/R34">Ancora paper</a> for an
explanation of the rationale behind this). The files we download from UCSC
are typically compressed with gzip. There is no need to decompress them,
because ceScan can read gzip-compressed files.</p>

<p>As an example, to detect CNEs between the current human and mouse genome
assemblies, the required alignment files can be obtained as follows:</p>
<pre>rsync -avzP \
  rsync://hgdownload.cse.ucsc.edu/goldenPath/hg18/vsMm9/axtNet/* \
  /export/downloads/ucsc/axtNet/hg18
rsync -avzP \
  rsync://hgdownload.cse.ucsc.edu/goldenPath/mm9/vsHg18/axtNet/* \
  /export/downloads/ucsc/axtNet/mm9</pre>

<p><em>David has some scripts that automate download of alignment and
annotation data. Those should perhaps be added to the repository (under
cne/scripts/cne_pipeline) and described here.</em></p>

<h3 id="Creating">2.2. Creating filter files</h3>

<p>Filer files list regions to be ignored in the scan. One filter file should
be created for each assembly. The current file for human assembly hg18 is
/export/data/CNEs/hg18/filters/filter_regions.hg18.bed, and files for other
assemblies are in corresponding locations. Filter files should adhere to the
three-column bed format. The filter files do not have to be sorted or
non-redundant (i.e. they may contain overlapping features).</p>

<p>We produce filter files from exon and repeat annotation downloaded from
UCSC. We maintain a list of annotation types that constitute the filter for
each genome at <a
href="http://ancora.genereg.net/filters.html">http://ancora.genereg.net/filters.html</a>.</p>

<p>Please refer to the section &ldquo;Obtaining genome annotations&rdquo;
below for information about how to download exon and repeat annotation from
UCSC and load it into a local MySQL database. Once the data is in a local
database, the script cne/scripts/cne_pipeline/make_filter.pl in the cne
package can be used to create a filter file from it. Run the script without
arguments for usage information. <em></em>The username and password that the
script should use to connect to the database must be declared in
MyPerlVars.pm (see above).<em>Todo: this script needs to be cleaned up a bit
and documented better here; possibly the wrapper script make_filter_batch.pl
should be integrated with it.</em> </p>

<h3 id="Scanning">2.3. Scanning for CNEs</h3>

<p>The cne package contains a Perl script
cne/scripts/cne_pipeline/detect_cnes.pl that we run to detect CNEs. The
simplest way to run the script is as follows:</p>
<pre>perl detect_cnes.pl all</pre>

<p>If executed in this way, the script will scan for CNEs for all pairwise
comparisons and thresholds listed in the configuration file comparisons.txt
located in the same directory as the script itself. Running the script with a
pair of assembly ids as arguments causes it to detect CNEs only for that
particular pairwise comparison (which must be listed in the configuration
file):</p>
<pre>perl detect_cnes.pl hg18 mm9</pre>

<p>If thresholds are also specified on the command line, the script will not
attempt to read a configuration file, but simply generate CNEs for the
indicated comparison and thresholds:</p>
<pre>perl detect_cnes.pl hg18 mm9 45,50 48,50 49,50</pre>

<p>Each threshold is specified as two values separated by a comma:
<em>identities</em>,<em>columns</em> - e.g. 45,50 for 45 (95%) identities
over 50 alignment columns. Any number of thresholds can be given. The
configuration file is formatted in the same manner; see the default file
comparisons.txt for an example. Some additional usage information can be
obtained by running the script without any arguments.</p>

<p>Briefly, the script does the following:</p>
<ul>
  <li>For each pair of genome assemblies <em>a</em> and <em>b</em> that are
    to be compared: 
    <ul>
      <li>Run ceScan to detect CNEs based on net alignments with <em>a</em>
        as reference.</li>
      <li>Run ceScan to detect CNEs based on net alignments with <em>b</em>
        as reference.</li>
      <li>Run the script merge_twoway_cne_sets.pl to combine the output from
        the two ceScan runs. In this process, any CNEs that overlap on both
        genomes are merged.</li>
    </ul>
  </li>
</ul>

<p>Output files are placed in the current working directory. The files are
named according the parameters of the comparison. For example, a search for
CNEs between assemblies hg18 and mm9 at a threshold of 49 identities over 50
columns generates a file named cne2w_hg18_mm9_49_50. Assembly ids are ordered
alphanumerically in the filenames. We currently keep these files under
/export/data/CNEs/pre-blatFilter/.</p>

<p>The output files contain one line for each CNE. There are nine columns:</p>
<ul>
  <li>Sequence name in first assembly</li>
  <li>Start coordinate in first assembly</li>
  <li>End coordinate in first assembly</li>
  <li>Sequence name in second assembly</li>
  <li>Start coordinate in second assembly</li>
  <li>End coordinate in second assembly</li>
  <li>Alignment orientation (+ or -)</li>
  <li>Percent identity (percentage of alignment columns with identities)</li>
  <li>CIGAR string, according to the DAS specification for alignments (see
    http://www.dasregistry.org/extension_alignment.jsp).</li>
</ul>

<p>The coordinate ranges are half-open, zero-based as in the bed format, so
that three-column bed files can be generated by cutting out columns 1-3 or
4-6.</p>

<h3 id="Removing">2.4. Removing unannotated repeats with BLAT</h3>

<p>This step has been implemented as a separate script because it takes long
to run. The goal of this step is to remove CNEs that correspond to repeated
sequence not annotated as such in the RepeatMasker track from UCSC. The
strategy is simple: use blat to align all CNEs back to the genome and remove
those having more than <em>N</em> hits with percent identity of at least
<em>I</em>. Based on some testing by David Fredman we use <em>I</em> = 90%
for all genomes. We use <em>N</em> = 4 for all genomes except the fish
genomes, for which we use <em>N</em> = 8. The largest family of duplicated
CNEs we have found is linked to the <em>IRX</em> genes and comprises 4
homologous CNEs in the human genome, motivating the chosen values of
<em>N</em> for land vertebrates. The two-fold higher value for fish is
motivated by the extra gene duplication that occurred in the teleost lineage.
We are currently using <em>N</em> = 4 for nematodes and flies as well,
although we have not explored CNE duplication in those genomes. Note that
<em>N</em> should not be set to anything lower than 2 because many assemblies
contain some artificially duplicated sequence (this means that <em>N</em>=4
may in fact be too strict for assemblies that may contain artificially
duplicated sequence from the <em>IRX</em> clusters, e.g. an <em>IRX</em>
cluster contig not assigned to a chromosome; we have not explored this
further).</p>

<p>The blat filter is implemented in the script
cne/scripts/cne_pipeline/blat_filter.pl. It can simply be run with the cne
files as arguments, e.g.:</p>
<pre>perl blat_filter.pl cne2w_hg18_mm9_49_50 cne cne2w_hg18_mm9_50_50</pre>

<p>Any number of CNE files can be given as arguments. For each input file, a
corresponding file with filtered CNEs will be created in the working
directory. This file will have the same name, but with the prefix cne2wBf
instead of cne2w. The format is also the same, except that four columns with
diagnostic information are added at the end (see the script for details).
Future versions of the script might not output these extra columns. We
currently keep the output files under /export/data/CNEs/blatFiltered.</p>

<p>The script reads the <em>N</em>-values for different assemblies from a
configuration file. By default, the script looks for a file named
blat_hit_cutoffs.txt located in the same directory as the script itself. Run
the script without any arguments for more information about how it can be
configured.</p>

<h3 id="Loading">2.5. Loading CNEs into the cne database</h3>

<p>The script cne/scripts/cne_pipeline/load_cnes.pl loads CNEs into a MySQL
database. For example, the following command loads two files with hg18-mm9
CNEs into the database cne:</p>
<pre>perl load_cnes.pl -d cne cne2wBf_hg18_mm9_49_50 cne2wBf_hg18_mm9_50_50</pre>

<p>Any number of CNE files can be listed after each other.</p>

<p>The username and password that the script should use to connect to the
database must be declared in MyPerlVars.pm (see above). Note that this MySQL
account must have write access to the database cne, or the script will
fail.</p>

<h2 id="Setting1">3. Setting up a genome browser</h2>

<p>Setting up a genome browser for a new assembly involves creating and
loading CNEs for the assembly, as described in the previous section, as well
as a number of additional steps detailed in this section.</p>

<h3 id="Creating1">3.1. Creating an annotation database for GBrowse</h3>

<p>Create a MySQL database to store the annotation (genes etc) that is to be
shown alongside the CNEs. For the main Ancora installation on shire, we name
these databases gbrowse_gff_&lt;assembly_id&gt;, e.g. gbrowse_gff_hg18 for
the latest human assembly. Make sure the user nobody@localhost has SELECT
permission on the database. If you don&rsquo;t know how to do this, refer to
the section on creating the CNE database above or to the documentation for
GBrowse.</p>

<h3 id="Obtaining1">3.2. Obtaining genome annotations</h3>

<p>This step usually involves more manual work than the other steps, because
annotations are available in diverse databases and in different and changing
formats. To keep things simple, we have tried to retrieve as much annotation
as possible from the UCSC Genome Browser database.</p>

<p>Begin by creating a local database to hold UCSC annotations for the
assembly, if one does not already exist. We have named these databases
UCSC_<em>id</em>, replacing <em>id</em> with the UCSC assembly id, e.g.
UCSC_hg18, UCSC_mm9.</p>

<p>The annotations available for download from UCSC can be browsed at <a
href="http://hgdownload.cse.ucsc.edu/downloads.html">http://hgdownload.cse.ucsc.edu/downloads.html</a>.
Annotations files can be downloaded in batch with rsync. E.g. the following
command gets all gap and RepeatMasker annotation for assembly hg18:</p>
<pre>rsync -avzP \
  rsync://hgdownload.cse.ucsc.edu/goldenPath/hg18/database/*_gap.*
  rsync://hgdownload.cse.ucsc.edu/goldenPath/hg18/database/*_rmsk.* .</pre>

<p>The UCSC annotations come as MySQL table dumps, with one table definition
file (ending in .sql) and one data file (ending in .txt.gz) for each table.
The following series of commands can be used to load a set of downloaded
files into a local MySQL database (named UCSC_hg18 in this example).</p>
<pre>mysql -p UCSC_hg18 &lt; *.sql</pre>
<pre>gunzip *.txt.gz</pre>
<pre>mysqlimport -Lp UCSC_hg18 *.txt</pre>

<p>Gap and RepeatMasker annotation, downloaded as in the example above, is
currently used in all Ancora browsers, so these two types of annotation
should always be downloaded if available. Which other annotation types are
included varies between the Ancora browsers. Table 1 below shows other
annotation from UCSC that we currently use and which corresponding tables
that need to be downloaded from UCSC.</p>

<table border="1" style="width: 100%">
  <caption><strong>Table 1. Annotation from UCSC used in Ancora
  browsers</strong></caption>
  <thead>
    <tr>
      <th>Annotation type</th>
      <th>UCSC tables</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Gaps</td>
      <td>*_gap</td>
    </tr>
    <tr>
      <td>RepeatMasker</td>
      <td>*_rmsk</td>
    </tr>
    <tr>
      <td>RefSeq genes</td>
      <td>refGene, refLink</td>
    </tr>
    <tr>
      <td>UCSC Known Genes</td>
      <td>knownGene, knownIsoforms, kgXref</td>
    </tr>
    <tr>
      <td>Aligned UCSC Known Gene peptides from other species</td>
      <td>blast*KG (for the target species)<br />
        kgXref (for the other species; should be loaded into a database for
        that species)</td>
    </tr>
    <tr>
      <td>ZFIN Genes</td>
      <td>vegaGene, vegaInfoZfish, refGene, refLink<br />
        (a file mapping ZFIN ids to Entrez gene ids is also needed, see usage
        information for ucsc2gff.pl)</td>
    </tr>
    <tr>
      <td>CpG islands</td>
      <td>cpgIslandExt</td>
    </tr>
    <tr>
      <td>ORegAnno Regulatory Elements</td>
      <td>oreganno, oregannoAttr</td>
    </tr>
  </tbody>
</table>

<p>The Ancora browsers also use annotation from sources other than UCSC. For
each of these sources, we download flat files as detailed in Table 2
below.</p>

<table border="1" style="width: 100%">
  <caption><strong>Table 2. Non-UCSC annotation used in Ancora
  browsers</strong></caption>
  <thead>
    <tr>
      <th>Annotation type</th>
      <th>Flat files</th>
      <th>Parsing script</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Ensembl genes</td>
      <td>Tab-separated text file obtained using BioMart. See usage message
        for ens2gff.pl for details.</td>
      <td>ens2gff.pl</td>
    </tr>
    <tr>
      <td>miRBase microRNAs</td>
      <td>GFF file obtained from
        http://microrna.sanger.ac.uk/sequences/ftp.shtml</td>
      <td>mirbase2gff.pl</td>
    </tr>
    <tr>
      <td>MGI genes</td>
      <td>Tab-delimited MGI coordinate file (currently named
        MGI_Coordinate.rpt) obtained from
        ftp://ftp.informatics.jax.org/pub/reports/index.html</td>
      <td>mgi2gff.pl</td>
    </tr>
    <tr>
      <td>FlyBase genes</td>
      <td>GFF file from ftp://ftp.flybase.net/. Currently used file is named
        dmel-all-r5.2.gff.gz.</td>
      <td>extract_genes_from_gff3.pl</td>
    </tr>
    <tr>
      <td>WormBase genes</td>
      <td>GFF file from
        ftp://ftp.wormbase.org/pub/wormbase/genomes/elegans/genome_feature_tables/GFF3<br
        />
        File mapping gene and transcript IDs to symbols from
        ftp://ftp.wormbase.org/pub/wormbase/genomes/elegans/annotations/gene_IDs</td>
      <td>extract_genes_from_gff3.pl</td>
    </tr>
    <tr>
      <td>RedFly CRMs</td>
      <td>GFF file from http://redfly.ccr.buffalo.edu/ (obtained by clicking
        &ldquo;Search&rdquo;, then &ldquo;Download all CRM&rdquo;)</td>
      <td>redfly2gff.pl</td>
    </tr>
    <tr>
      <td>RedFly TFBSs</td>
      <td>GFF file from http://redfly.ccr.buffalo.edu/ (obtained by clicking
        &ldquo;Search&rdquo;, then &ldquo;Download all TFBS&rdquo;)</td>
      <td>redfly2gff.pl</td>
    </tr>
    <tr>
      <td>Syteny blocks</td>
      <td>Output from join_nets.pl (for vertebrates) or Drosophila synteny
        pipeline. To be documented.</td>
      <td>synteny2gff.pl</td>
    </tr>
    <tr>
      <td>Target genes</td>
      <td>File cne/ancora/gbrowse_db/targets/ancora_targetGenes.hg18.bed in
        the cne package</td>
      <td>targets2gff.pl</td>
    </tr>
  </tbody>
</table>

<p>Note that when downloading coordinate files such as these it is essential
to check what assembly version coordinates refer to. If the version is not
the same as that to be shown in Ancora, coordinates can be converted between
assemblies using the <a
href="http://genome.ucsc.edu/cgi-bin/hgLiftOver">liftOver tool</a> on the
UCSC web site. For the MGI genes, the scripts mgi2bed.pl and bed2mgi.pl in
directory cne/ancora/gbrowse_db/mgi in the cne package are useful for
converting an MGI coordinate file to bed format for coordinate conversion,
and then converting the liftOver output back to MGI format.</p>

<h3 id="Creating2">3.3. Creating and loading GFF files</h3>

<p>The annotations in the UCSC MySQL tables and the flat files must be
converted to a GFF format suitable for import into the GBrowse annotation
database. A number of scripts that carry out those conversions are available
in the directory cne/ancora/gbrowse_db in the cne package. The script
ucsc2gff.pl extracts data from UCSC annotation tables stored in a MySQL
database and outputs GFF format suitable for import into the GBrowse
database. The scripts that process flat files to generate GBrowse GFF files
are listed in Table 2 above.</p>

<p>The script ucsc2gff.pl can also read genome assembly information from 2bit
files and output a GFF feature for each sequence in the assembly (the
chromosomes, or supercontigs if the genome is not assembled into
chromosomes). These GFF lines are also needed by GBrowse.</p>

<p>In the same directory as the Perl conversion scripts, there is a bash
script make_gff.sh that runs the conversion scripts to generate GFF files for
all assemblies currently in Ancora. The script creates GFF files named by
assembly id (e.g. hg18.gff, mm9.gff). It is recommended carry out the
conversions by running this bash script instead of running the Perl scripts
directly, as it also provides a record of which annotation is being used for
each assembly. The bash script does not require any arguments, but it expects
the UCSC databases with relevant annotation tables to be available on the
local host, as well as relevant flat files to be present in the correct
subdirectories. Please see the script for where to place the flat files so
that it can find them.</p>

<p>When GFF files have been created they can be loaded into GBrowse
annotation databases using the script load_gff.sh. The script does not take
any arguments, but prompts for a MySQL username and password. NOTE: use this
script with caution as it erases all data stored in the GBrowse annotation
databases before it loads the GFF files.</p>

<h3 id="Adding">3.4. Adding information about the assembly to the CNE
database</h3>

<p>Ancora needs some information about each assembly. As described above,
this information is stored in a table called assembly in the database cne. If
the assembly you are setting up a browser for is not already present in this
table, you need to execute an SQL INSERT statement (see the MySQL manual) to
add a row describing the assembly. Most of the fields in the table are
self-explanatory. The following may not be:</p>
<ul>
  <li>ensembl_ver &ndash; Currently only used by the DAS service. Specifies
    which Ensembl version that DAS tracks should be added to. Leave blank for
    the most recent Ensembl release, or add a version in the form of an
    archive name (e.g. &ldquo;apr2007&rdquo;) for an older release.</li>
  <li>default_ensembl_loc &ndash; Currently only used by the DAS service.
    Specifies which location the user should be taken to in Ensembl when
    tracks are added. Specify as an Ensembl location string (e.g.
    &ldquo;7:8541098-8656549&rdquo;).</li>
  <li>ucsc_db &ndash; Deprecated, so can be left blank. Earlier releases of
    Ancora required an UCSC annotation database to be present for some
    assemblies.</li>
</ul>

<h3 id="Creating3">3.5. Creating a configuration file for GBrowse</h3>

<p>GBrowse requires a configuration file for each assembly. We keep the live
versions of these files under /opt/www/ancora/conf/gbrowse.conf. The files
are also present in the cne repository, under cne/ancora/conf/gbrowse.conf.
The files are named by assembly id and have the suffix .conf (e.g. hg18.conf
for the current human assembly). The configuration files are fairly complex,
so the easiest way to set up a file for a new assembly is to copy one written
for a different assembly and modify it. The file hg18.conf begins with a few
comment lines listing what should be changed when adapting a file for a new
assembly. It is typically best to start from a file for a different assembly
for the same species, if available, and otherwise from a file for a similar
species. The reason for this is that the types of annotation available
differs between species and some parameters, in particular the semantic zoom
levels, depend on genome compactness. The general format of GBrowse
configuration files is well described in the <a
href="http://www.gmod.org/wiki/index.php/CONFIGURE_HOWTO">GBrowse
configuration HOWTO</a>, which should be read before editing a configuration
file.</p>

<p>The configuration of plugins that display CNE locations and densities
deserves a more detailed description. These plugins are configured in three
places in the configuration file:</p>

<p>(1) The line starting with &ldquo;plugins =&rdquo; specifies what plugins
are to be loaded. For example, if CNEs from three comparisons are to be
shown, this should be specified as:</p>
<pre>plugins = CNEPlotInstance1 CNEPlotInstance2 CNEPlotInstance3</pre>

<p>Each of the plugins listed on this line corresponds to a file in the
subdirectory plugins under the configuration directory. The files
CNEPlotInstance*.pm in this directory are identical subclasses of the same
base class Bio::Graphics::Browser::Plugin::CNEPlot. This is a trick to be
able to load several instances of the same plugin. There are currently nine
CNE plugin files (up to CNEPlotInstance9.pm) in the plugins directory, so up
to nine comparisons can be shown in any Ancora browser. If you wish to show
more comparisons, simply add additional plugin files by copying one of the
existing files and changing number on the first line in the new file.</p>

<p>(2) Parameters for each plugin are specified in the section following the
comment</p>
<pre># Plugin config</pre>

<p>One parameter must be specified for each plugin:</p>
<ul>
  <li>asm_id &ndash; Id for the other assembly in the comparison (e.g.
  hg18)</li>
</ul>

<p>The following general parameters can also be specified:</p>
<ul>
  <li>asm_name &ndash; Name to be displayed for the other assembly in the
    comparison. Usually the common name for the species (e.g. Human).
    Defaults to asm_id.</li>
  <li>window_size &ndash; Window size, in kb, to use for density
    computations. Defaults to 50 kb.</li>
  <li>max_score &ndash; Height of density plot. Defaults to 20.</li>
  <li>nr_cne_sets &ndash; Number of CNE sets from this comparison that the
    user can define and simultaneously view. Defaults to 3.</li>
  <li>nr_graphs &ndash; Separate densities based on chromosome in other
    genome into this number of graphs. Defaults to 1 (meaning don&rsquo;t
    separate).</li>
  <li>rank_un_low &ndash; Give low priority to poorly assembled sequence and
    alternative haplotypes when separating densities. Defaults to 0 (no). Set
    this option to a true value to activate it.</li>
</ul>

<p>In addition, the following parameters can also be used to define CNE sets,
replacing <em>i</em> with a set number (1,2,3,&hellip;):</p>
<ul>
  <li>min_cne_id<em>i</em> - Similarity threshold for the CNE set. This
    parameter specifies which CNE set to retrieve from the database, so it
    must correspond to a CNE set present in the database. Specified as:
    <em>percent-identity</em>/<em>columns</em>. E.g.: &ldquo;0.98/50&rdquo;
    for 98% identity over 50 alignment columns. Default: undefined.</li>
  <li>min_cne_len<em>i</em> - Required shortest length for CNEs, in bp.
    Default: 50 bp.</li>
  <li>show_cnes<em>i</em> - Show CNE locations. Default: 0 (no). Set to a
    true value to activate.</li>
  <li>show_graph<em>i</em> - Show density plot. Default: 0 (no). Set to a
    true value to activate.</li>
</ul>

<p>All of the above parameters except asm_id and asm_name can be modified by
the user via the web interface. GBrowse uses cookies to save user
settings.</p>

<p>(3) At the end of the configuration file, there is a track definition for
each CNE plugin, e.g.:</p>
<pre>[plugin:Mouse HCNEs]
citation = HCNEs conserved in mouse.
See &lt;a href="/methods.html"&gt;Methods&lt;/a&gt; for a description of how these were detected.</pre>

<p>The order in which the track definitions are listed in the configuration
file specifies the default order of the CNE tracks in the browser. This order
can be changed by the user, who can also restore the default order by
clicking the red &ldquo;[Reset]&rdquo; link in the genome browser.</p>

<h3 id="Making">3.6. Making download files</h3>

<p>Ancora has a download section where files with CNE locations (bed format)
and CNE densities (wig format) can be downloaded. The files are located under
/opt/www/ancora/html/downloads. Download files can be created using the Perl
script cne/scripts/cne_pipeline/make_download_files.pl in the cne package.
Run the script without arguments for usage information. There is a wrapper
bash script make_download_files.sh that runs the Perl script to create files
for the assemblies currently shown in Ancora. It is therefore recommended to
use this bash script for updating the Ancora download files. The bash script
does not require any arguments.</p>

<p>The output files are placed in subdirectories created under the directory
from which the scripts are run. The directory structure created is identical
to that in the Ancora download section, so that the directories created by
the scripts simply can be moved to the Ancora download directory
/opt/www/ancora/html/downloads after checking that the files seem OK.</p>

<h2 id="HTML">4. The HTML pages and CGI scripts</h2>

<p>The HTML pages are located under cne/ancora/html in the cne package. These
files should be identical to the live HTML pages that we keep under
/opt/www/ancora/html. The style of the pages is defined in the stylesheet
cne/ancora/html/main.css. A stylesheet for GBrowse
(cne/ancora/html/gbrowse/gbrowse-cne.css) is also part of the cne package, to
give GBrowse a look and feel similar to the Ancora pages.</p>

<p>There are three CGI scripts for Ancora, located under
/opt/www/ancora/cgi-bin (live version) and under cne/ancora/cgi-bin in the
cne package:</p>
<ul>
  <li>cne_color_legend - Shows available color legends for chromosome colors.
    Ancora runs this script in a popup window opened from the CNE track
    configuration page.</li>
  <li>cne_das_loader - Handles addition of DAS tracks to Ensembl.</li>
  <li>cne_details - Shows details of CNEs or synteny block. Invoked if user
    clicks on one of these features in the genome browser.</li>
</ul>

<h2 id="service">5. The DAS service</h2>

<p>Ancora uses ProServer (VERSION?), a DAS server implemented in Perl and
available at http://www.sanger.ac.uk/Software/analysis/proserver/. ProServer
is installed under /opt/www/ancora/Bio-Das-ProServer. The directory
/opt/www/ancora/Bio-Das-ProServer/ancora contains Ancora-specific files
related to the DAS server:</p>
<ul>
  <li>ancora_das - Init script. A copy should be present in /etc/init.d.</li>
  <li>start_proserver - Script called by ancora_das to start ProServer.</li>
  <li>proserver.ini - ProServer config file.</li>
</ul>

<p>The same files are present in the cne package, under
cne/ancora/Bio-Das-ProServer/ancora.</p>

<p>To start ProServer do:</p>
<pre>sudo /etc/init.d/ancora_das start</pre>

<p>To stop it:</p>
<pre>sudo /etc/init.d/ancora_das stop</pre>

<p>To stop and start it:</p>
<pre>sudo /etc/init.d/ancora_das restart</pre>

<p>The config file proserver.ini defines which DAS sources are available.
ProServer retrieves CNE locations from the same database (cne on localhost)
as the Ancora genome browser. This is handled by three modules in the cne
package: Bio::Das::ProServer::SourceAdaptor::MyCNEAdpator,
Bio::Das::ProServer::SourceAdaptor::MyCNEDensityAdpator and
Bio::Das::ProServer::SourceAdaptor::Transport::MyCNEDatabase. CNE densities
are computed from CNE locations on the fly, as done by the genome browser.</p>
<hr />

<p></p>

<p></p>

<p></p>
</body>
</html>
